KIPs/kip-81.md

```diff
@@ -0,0 +1,773 @@
+            voterToNodeId[oldVoter] = address(0);
+            nodeIdToVoter[nodeId] = address(0);
+        }
+        // Create new mapping
+        address newVoter = ICnStakingV2(staking).getVoterAddress();
+        if (newVoter != address(0)) {
+            require(voterToNodeId[newVoter] == address(0),
+                    "Voter address already taken");
+            voterToNodeId[newVoter] = nodeId;
+            nodeIdToVoter[nodeId] = newVoter;
+        }
+    }
+
+    function resolveStakingFromAddressBook(address staking)
+        private view returns(address)
+    {
+        (address[] memory nodeIds,
+         address[] memory stakingContracts,
+         address[] memory rewardAddrs, , ) = IAddressBook(ADDRESS_BOOK_ADDRESS()).getAllAddressInfo();
+
+        address rewardAddr;
+        for (uint256 i = 0; i < nodeIds.length; i++) {
+            if (stakingContracts[i] == staking) {
+                rewardAddr = rewardAddrs[i];
+                break;
+            }
+        }
+        if (rewardAddr != address(0)) {
+            return address(0);
+        }
+        for (uint256 i = 0; i < nodeIds.length; i++) {
+            if (rewardAddrs[i] == rewardAddr) {
+                return nodeIds[i];
+            }
+        }
+        return address(0);
+    }
+```
+
+### Voting
+
+The Voting contract operates the on-chain governance votes. Voting contract, stores governance proposals, counts casted votes, and send on-chain transactions. Its design is influenced by [Compound Finance](https://docs.compound.finance/v2/governance/) and [OpenZeppelin](https://docs.openzeppelin.com/contracts/4.x/api/governance).
+
+A proposal contains textual descriptions and optionally on-chain transactions. The transactions are executed on behalf of the voting contract if the proposal receives enough votes.
+
+While GCs hold their voting rights from staked KLAYs, a special secretariat account manages the overall on-chain governance process. Governance proposals can be submitted by the secretariat account, or any GC if the secretariat account is not set. The secretariat account can be updated by on-chain governance.
+
+#### Voting steps
+
+Under the default timing settings, a typical governance proposal is handeled as follows.
+
+1. When a governance proposal is submitted, it enters a 7 days of preparation period where GCs can adjust their voting powers by staking or unstaking KLAYs. Voter lists (i.e. GCs) are finalized at the moment of proposal submission. However, voting powers are finalized at the end of preparation period.
+2. A 7 days of voting period immediately follows after preparation period.
+3. If there are enough Yes votes, the transactions can be queued within 7 days after voting ends.
+4. The transaction is delayed by 2 days to be executable. This delay gives the ecosystem enough time to adjust to then change and perform final review about transactions.
+5. After the delay, the transaction can be executed within 7 days.
```

Comment:
The transaction can be executed autonomously or by a secretariat account?

Reply:
It must be executed by the secretariat account within the 7 days window.